src/oscillators/sa5x_oscillator.c

```diff
@@ -0,0 +1,381 @@
+		memset(answer_str, 0, rbytes);
+		return -1;
+	}
+
+	return rbytes;
+}
+
+static int sa5x_oscillator_read_intval(int *val)
+{
+	return sscanf(answer_str, "[=%d]\r\n", val);
+}
+
+static int sa5x_oscillator_get_attributes(struct oscillator *oscillator, struct sa5x_attributes *a,
+										  unsigned int attributes_mask)
+{
+	struct sa5x_oscillator *sa5x;
+	sa5x = container_of(oscillator, struct sa5x_oscillator, oscillator);
+	int err;
+
+	if (attributes_mask & (ATTR_CTRL|ATTR_STATUS) && !a) {
+		log_error("scillator_get_attributes no structure provided");
+		return -1;
+	}
+
+	if (attributes_mask & ATTR_FW_SERIAL) {
+		err = sa5x_oscillator_cmd(sa5x, CMD_SWVER, sizeof(CMD_SWVER));
+		if (err > 0) {
+			sscanf(answer_str, "[=%20[^,],", sa5x->version);
+			memset(answer_str, 0, err);
+		}
+
+		err = sa5x_oscillator_cmd(sa5x, CMD_SERIAL, sizeof(CMD_SERIAL));
+		if (err > 0) {
+			sscanf(answer_str, "[=%12s]\r\n", sa5x->serial);
+			memset(answer_str, 0, err);
+		}
+		return 0;
+	}
+
+	if (attributes_mask & ATTR_CTRL) {
+		err = sa5x_oscillator_cmd(sa5x, CMD_GET_LOCKED, sizeof(CMD_GET_LOCKED));
+		if (err > 0) {
+			int val;
+			if (sa5x_oscillator_read_intval(&val) == 1) {
+				a->locked = val;
+			}
+			memset(answer_str, 0, err);
+		} else {
+			// this is the only parameter that we depend on
+			return err;
+		}
+
+		err = sa5x_oscillator_cmd(sa5x, CMD_GET_PHASE, sizeof(CMD_GET_PHASE));
+		if (err > 0) {
+			int val;
+			if (sa5x_oscillator_read_intval(&val) == 1) {
+				a->phaseoffset = val;
+			}
+			memset(answer_str, 0, err);
+		}
+
+		err = sa5x_oscillator_cmd(sa5x, CMD_GET_LASTCORRECTION, sizeof(CMD_GET_LASTCORRECTION));
+		if (err > 0) {
+			int val;
+			if (sa5x_oscillator_read_intval(&val) == 1) {
+				a->lastcorrection = val;
+			}
+			memset(answer_str, 0, err);
+		}
+	}
+
+	if (attributes_mask & ATTR_STATUS) {
+		err = sa5x_oscillator_cmd(sa5x, CMD_GET_TEMPERATURE, sizeof(CMD_GET_TEMPERATURE));
+		if (err > 0) {
+			int val;
+			if (sa5x_oscillator_read_intval(&val) == 1) {
+				a->temperature = val;
+			}
+			memset(answer_str, 0, err);
+		} else {
+			// this is the only parameter that we depend on
+			return err;
+		}
+
+		err = sa5x_oscillator_cmd(sa5x, CMD_GET_ANALOG_TUNING_ENABLED, sizeof(CMD_GET_ANALOG_TUNING_ENABLED));
+		if (err > 0) {
+			int val;
+			if (sa5x_oscillator_read_intval(&val) == 1) {
+				a->analogtuningon = val;
+			}
+			memset(answer_str, 0, err);
+		}
+
+		err = sa5x_oscillator_cmd(sa5x, CMD_GET_ANALOG_TUNING, sizeof(CMD_GET_ANALOG_TUNING));
+		if (err > 0) {
+			int val;
+			if (sa5x_oscillator_read_intval(&val) == 1) {
+				a->analogtuning = val;
+			}
+			memset(answer_str, 0, err);
+		}
+
+		err = sa5x_oscillator_cmd(sa5x, CMD_GET_DIGITAL_TUNING, sizeof(CMD_GET_DIGITAL_TUNING));
+		if (err > 0) {
+			int val;
+			if (sa5x_oscillator_read_intval(&val) == 1) {
+				a->digitaltuning = val;
+			}
+			memset(answer_str, 0, err);
+		}
```

Comment:
@vvfedorenko The last 8 lines are repeated multiple times, don't you think we should factorize this ?

Reply:
yes, Charles, you are right. Updated the diff